core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java

```diff
@@ -224,62 +140,97 @@ private static void failLateSubscriber(SubscriptionImpl subscription, Subscriber
     public void abort() {
         final SubscriptionImpl currentSubscription = subscription;
         if (currentSubscription != null) {
-            currentSubscription.abort();
+            cancelOrAbort(false);
             return;
         }
 
         final SubscriptionImpl newSubscription = new SubscriptionImpl(
                 this, AbortingSubscriber.get(), null, false);
         if (subscriptionUpdater.compareAndSet(this, null, newSubscription)) {
             // We don't need to invoke onSubscribe() for AbortingSubscriber because it's just a placeholder.
-            newSubscription.invokedOnSubscribe = true;
-            newSubscription.abort();
-        } else {
-            subscription.abort();
+            invokedOnSubscribe = true;
         }
+        cancelOrAbort(false);
     }
 
     @Override
-    public boolean write(T obj) {
-        requireNonNull(obj, "obj");
-        if (obj instanceof ReferenceCounted) {
-            ((ReferenceCounted) obj).touch();
-            if (!(obj instanceof ByteBufHolder) && !(obj instanceof ByteBuf)) {
-                throw new IllegalArgumentException(
-                        "can't publish a ReferenceCounted that's not a ByteBuf or a ByteBufHolder: " + obj);
-            }
-        }
-
-        if (!isOpen()) {
-            ReferenceCountUtil.safeRelease(obj);
-            return false;
-        }
-
+    void addObject(T obj) {
         wroteAny = true;
-        pushObject(obj);
-        return true;
+        addObjectOrEvent(obj);
     }
 
     @Override
-    public boolean write(Supplier<? extends T> supplier) {
-        return write(supplier.get());
+    long demand() {
+        return demand;
     }
 
     @Override
-    public CompletableFuture<Void> onDemand(Runnable task) {
-        requireNonNull(task, "task");
+    void request(long n) {
+        for (;;) {
+            final long oldDemand = demand;
+            final long newDemand;
+            if (oldDemand >= Long.MAX_VALUE - n) {
+                newDemand = Long.MAX_VALUE;
+            } else {
+                newDemand = oldDemand + n;
+            }
 
-        final AwaitDemandFuture f = new AwaitDemandFuture();
-        if (!isOpen()) {
-            f.completeExceptionally(ClosedPublisherException.get());
-            return f;
+            if (demandUpdater.compareAndSet(this, oldDemand, newDemand)) {
+                if (oldDemand == 0) {
+                    notifySubscriber();
+                }
+                break;
+            }
         }
+    }
 
-        pushObject(f);
-        return f.thenRun(task);
+    @Override
+    void cancel() {
+        cancelOrAbort(true);
     }
 
-    private void pushObject(Object obj) {
+    private void cancelOrAbort(boolean cancel) {
+        if (setState(State.OPEN, State.CLEANUP)) {
```

Comment:
By the way, this code is the major change for `DefaultStreamMessage` in this PR. I noticed that `cancelRequested` was not set when entering this block - I couldn't think of a reason for that behavior, so I set it always when `Subscription.cancel` is called now. Also, `cancelRequested` got overwritten by `abort`, which also didn't seem correct to me. Hopefully this new flow makes sense but let's look at it carefully.

Reply:
Seems OK to me. To be honest, I'm not sure why such a field is necessary at all. I don't remember why it was needed either. It just looks cleaner now.